engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/BaseObjectUpdateByOperator.java

```diff
@@ -5,78 +5,92 @@
  */
 package io.deephaven.engine.table.impl.updateby.internal;
 
+import io.deephaven.engine.table.impl.util.ChunkUtils;
+
 import io.deephaven.chunk.Chunk;
-import io.deephaven.chunk.IntChunk;
 import io.deephaven.chunk.LongChunk;
-import io.deephaven.chunk.attributes.ChunkLengths;
-import io.deephaven.chunk.attributes.ChunkPositions;
+import io.deephaven.chunk.WritableObjectChunk;
 import io.deephaven.chunk.attributes.Values;
-import io.deephaven.chunk.sized.SizedObjectChunk;
-import io.deephaven.engine.rowset.*;
-import io.deephaven.engine.rowset.chunkattributes.OrderedRowKeys;
-import io.deephaven.engine.rowset.chunkattributes.RowKeys;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.rowset.RowSet;
 import io.deephaven.engine.table.*;
-import io.deephaven.engine.table.impl.UpdateBy;
-import io.deephaven.engine.table.impl.UpdateByOperator;
 import io.deephaven.engine.table.impl.sources.*;
+import io.deephaven.engine.table.impl.updateby.UpdateByCumulativeOperator;
 import io.deephaven.engine.table.impl.util.RowRedirection;
-import io.deephaven.engine.table.impl.util.SizedSafeCloseable;
-import io.deephaven.util.QueryConstants;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.Collections;
 import java.util.Map;
 
 import static io.deephaven.engine.rowset.RowSequence.NULL_ROW_KEY;
+import static io.deephaven.util.QueryConstants.*;
 
-public abstract class BaseObjectUpdateByOperator<T> implements UpdateByOperator {
+public abstract class BaseObjectUpdateByOperator<T> extends UpdateByCumulativeOperator {
     protected final WritableColumnSource<T> outputSource;
     protected final WritableColumnSource<T> maybeInnerSource;
-    protected final MatchPair pair;
-    protected final String[] affectingColumns;
-    protected final boolean isRedirected;
-
-    protected ObjectArraySource<T> bucketLastVal;
-
-    /** These are only used in grouped operations */
-    protected T singletonVal;
-    protected long singletonGroup = QueryConstants.NULL_LONG;
-
-    protected boolean initialized = false;
 
     // region extra-fields
     private final Class<T> colType;
```

Comment:
You will probably need to introduce component type, too, at least for rolling group (which is the windowed operators, I know).